### PR TITLE
Implement rechecking of redactions for room versions v3

### PR DESCRIPTION
**Pull request by erikjohnston**
This starts implementing support for V3 rooms, while keeping V3 disabled

### CODE REVIEW

File: synapse/api/constants.py [hunk 1]
```diff
@@ -104,6 +104,7 @@ class ThirdPartyEntityKind(object):
 class RoomVersions(object):
     V1 = "1"
     V2 = "2"
+    V3 = "3"  # Not currently fully supported, so
```
Comment (richvdh) [inline synapse/api/constants.py]: so!

**Comment by erikjohnston** [comment]
> erm: do we not need to check that the sender of a redaction matches the original sender when we pull the original event out of the database?

Argh, bollocks

File: synapse/storage/events_worker.py [hunk 2]
```diff
@@ -443,6 +466,19 @@ def _get_event_from_row(self, internal_metadata, js, redacted,
...
+                    # rechecked if we didn't have the redacted event at the
+                    # time, so we recheck on read instead.
+                    if because.internal_metadata.need_to_check_redaction():
+                        expected_domain = get_domain_from_id(original_ev.sender)
+                        if get_domain_from_id(because.sender) == expected_domain:
+                            # This redaction event is allowed. Mark as not needing a
+                            # recheck.
+                            because.internal_metadata.recheck_redaction = False
+                        else:
+                            # Senders don't match, so the event is actually redacted
```
Comment (erikjohnston) [inline synapse/storage/events_worker.py]:
```suggestion
                            # Senders don't match, so the event isn't actually redacted
```

**Comment by erikjohnston** [comment]
sytest looks to be unhappy that i killed off VDH_TEST

File: synapse/events/__init__.py [hunk 3]
```diff
@@ -63,6 +63,18 @@ def get_send_on_behalf_of(self):
         return getattr(self, "send_on_behalf_of", None)
 
     def need_to_check_redaction(self):
+        """Whether the redaction event needs to be rechecked when fetching
+        from the database.
+
+        Starting in room v3 redaction events are accepted up front, and later
+        checked to see if the redacter and redactee's domains match.
+
+        If the sender of the redaction event is allowed to redact due to auth
```
Comment (richvdh) [inline synapse/events/__init__.py]: "is allowed to redact any event"

**Pull request merged**